pkg/controllers/cloud/networkpolicy.go

```diff
@@ -872,47 +745,204 @@ func (a *appliedToSecurityGroup) updateRules(r *NetworkPolicyReconciler) error {
-func (a *appliedToSecurityGroup) computeRules(nps []interface{}) ([]*securitygroup.IngressRule, []*securitygroup.EgressRule) {
-	irules := make([]*securitygroup.IngressRule, 0)
-	erules := make([]*securitygroup.EgressRule, 0)
+func (a *appliedToSecurityGroup) clearMember(r *NetworkPolicyReconciler) {
+	if a.hasMembers {
+		r.Log.V(1).Info("AppliedToSecurityGroup clear members with no rules", "Name", a.id.Name)
+		ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupMembers(&a.id, nil, false)
+		go func() {
+			err := <-ch
+			r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationClearMembers, err: err}
+		}()
+		return
+	}
+	// No need to update appliedToSecurityGroup with no members.
+	r.Log.V(1).Info("AppliedToSecurityGroup clear rules", "Name", a.id.Name)
+	a.hasRules = false
+}
+
+// combineRules convert and combine all rules from given anps to securitygroup.CloudRule.
+func (a *appliedToSecurityGroup) combineRules(nps []interface{}) (map[string]*securitygroup.CloudRule, []*securitygroup.CloudRule) {
+	ruleMap := make(map[string]*securitygroup.CloudRule, 0)
+	ruleSlice := make([]*securitygroup.CloudRule, 0)
 	for _, i := range nps {
 		np := i.(*networkPolicy)
+		npName := np.getNamespacedName()
 		if !np.rulesReady {
 			return nil, nil
 		}
-		irules = append(irules, deepcopy.Copy(np.ingressRules).([]*securitygroup.IngressRule)...)
-		erules = append(erules, deepcopy.Copy(np.egressRules).([]*securitygroup.EgressRule)...)
+		for _, r := range np.ingressRules {
+			rule := &securitygroup.CloudRule{
+				Rule:          deepcopy.Copy(r).(*securitygroup.IngressRule),
+				NetworkPolicy: npName,
+				AppliedToGrp:  a.id.String(),
+			}
+			ruleMap[rule.GetUUID()] = rule
+			ruleSlice = append(ruleSlice, rule)
+		}
+		for _, r := range np.egressRules {
+			rule := &securitygroup.CloudRule{
+				Rule:          deepcopy.Copy(r).(*securitygroup.EgressRule),
+				NetworkPolicy: npName,
+				AppliedToGrp:  a.id.String(),
+			}
+			ruleMap[rule.GetUUID()] = rule
+			ruleSlice = append(ruleSlice, rule)
+		}
+	}
+	return ruleMap, ruleSlice
+}
+
+// computeRules compute the rule update delta of an anp.
+func (a *appliedToSecurityGroup) computeRules(r *NetworkPolicyReconciler, rules map[string]*securitygroup.CloudRule,
+	npName string) ([]*securitygroup.CloudRule, []*securitygroup.CloudRule, error) {
+	addRules := make([]*securitygroup.CloudRule, 0)
+	removeRules := make([]*securitygroup.CloudRule, 0)
+
+	currentRules, err := r.cloudRuleIndexer.ByIndex(cloudRuleIndexerByAppliedToGrp, a.id.String())
+	currentRuleMap := make(map[string]*securitygroup.CloudRule)
+	if err != nil {
+		r.Log.Error(err, "get cloudRule indexer", "sg", a.id.String())
+		return nil, nil, err
+	}
+	for _, obj := range currentRules {
+		currentRule := obj.(*securitygroup.CloudRule)
+		ruleUUID := currentRule.GetUUID()
+		currentRuleMap[ruleUUID] = currentRule
+
+		if currentRule.NetworkPolicy == "" {
```

Comment:
Find a better place to update the rule indexer

Reply:
Moved to a separate function, before ruleCompute happens